File: plugins/plugin-shell/rust/src/providers/shell_history.rs
use crate::{Provider, ProviderResult, ShellService};
use async_trait::async_trait;
use serde_json::{json, Value};
use std::time::{Duration, UNIX_EPOCH};

const MAX_OUTPUT_LENGTH: usize = 8000;
const TRUNCATE_SEGMENT_LENGTH: usize = 4000;

pub struct ShellHistoryProvider;

fn format_timestamp(timestamp: f64) -> String {
    let duration = Duration::from_secs_f64(timestamp);
    let _datetime = UNIX_EPOCH + duration;
    // Format as simple ISO-like string
    format!("{:.0}", timestamp)
}

#[async_trait]
impl Provider for ShellHistoryProvider {
    fn name(&self) -> &str {
        "SHELL_HISTORY"
    }

    fn description(&self) -> &str {
        "Provides recent shell command history, current working directory, \
         and file operations within the restricted environment"
    }

    fn position(&self) -> i32 {
        99
    }

    async fn get(
        &self,
        message: &Value,
        _state: &Value,
        service: Option<&ShellService>,
    ) -> ProviderResult {
        let service = match service {
            Some(s) => s,
            None => {
                return ProviderResult {
                    values: json!({
                        "shellHistory": "Shell service is not available",
                        "currentWorkingDirectory": "N/A",
                        "allowedDirectory": "N/A",
                    }),
                    text: "# Shell Status\n\nShell service is not available".to_string(),
                    data: json!({
                        "historyCount": 0,
                        "cwd": "N/A",
                        "allowedDir": "N/A",
                    }),
                };
            }
        };

        let conversation_id = message
            .get("room_id")
            .and_then(|r| r.as_str())
            .or_else(|| message.get("agent_id").and_then(|a| a.as_str()))
            .unwrap_or("default");

        let history = service.get_command_history(conversation_id, Some(10));
        let cwd = service.get_current_directory(None);
        let allowed_dir = service.get_allowed_directory();

        let history_text = if history.is_empty() {
            "No commands in history.".to_string()
        } else {
            history
                .iter()
                .map(|entry| {
                    let mut entry_str = format!(
                        "[{}] {}> {}",
                        format_timestamp(entry.timestamp),
                        entry.working_directory,
                        entry.command
                    );

                    if !entry.stdout.is_empty() {
                        let stdout = if entry.stdout.len() > MAX_OUTPUT_LENGTH {
                            format!(
                                "{}\n  ... [TRUNCATED] ...\n  {}",
                                &entry.stdout[..TRUNCATE_SEGMENT_LENGTH],
                                &entry.stdout[entry.stdout.len() - TRUNCATE_SEGMENT_LENGTH..]
                            )
                        } else {
                            entry.stdout.clone()
                        };
                        entry_str.push_str(&format!("\n  Output: {}", stdout));
                    }

                    if !entry.stderr.is_empty() {
                        let stderr = if entry.stderr.len() > MAX_OUTPUT_LENGTH {
                            format!(
                                "{}\n  ... [TRUNCATED] ...\n  {}",
                                &entry.stderr[..TRUNCATE_SEGMENT_LENGTH],
                                &entry.stderr[entry.stderr.len() - TRUNCATE_SEGMENT_LENGTH..]
                            )
                        } else {
                            entry.stderr.clone()
                        };
                        entry_str.push_str(&format!("\n  Error: {}", stderr));
                    }

                    if let Some(exit_code) = entry.exit_code {
                        entry_str.push_str(&format!("\n  Exit Code: {}", exit_code));
                    }

                    if let Some(ref file_ops) = entry.file_operations {
                        if !file_ops.is_empty() {
                            entry_str.push_str("\n  File Operations:");
                            for op in file_ops {
                                if let Some(ref secondary) = op.secondary_target {
                                    entry_str.push_str(&format!(
                                        "\n    - {:?}: {} → {}",
                                        op.op_type, op.target, secondary
                                    ));
                                } else {
                                    entry_str.push_str(&format!(
                                        "\n    - {:?}: {}",
                                        op.op_type, op.target
                                    ));
                                }
                            }
                        }
                    }

                    entry_str
                })
                .collect::<Vec<_>>()
                .join("\n\n")
        };

        let recent_file_ops: Vec<_> = history
            .iter()
            .filter_map(|e| e.file_operations.as_ref())
            .flat_map(|ops| ops.iter())
            .take(5)
            .collect();

        let file_ops_text = if recent_file_ops.is_empty() {
            String::new()
        } else {
            let ops_str = recent_file_ops
                .iter()
                .map(|op| {
                    if let Some(ref secondary) = op.secondary_target {
                        format!("- {:?}: {} → {}", op.op_type, op.target, secondary)
                    } else {
                        format!("- {:?}: {}", op.op_type, op.target)
                    }
                })
                .collect::<Vec<_>>()
                .join("\n");
            format!("\n\n# Recent File Operations\n\n{}", ops_str)
        };

        let cwd_str = cwd.display().to_string();
        let allowed_dir_str = allowed_dir.display().to_string();

        let text = format!(
            "Current Directory: {}\nAllowed Directory: {}\n\n# Shell History (Last 10)\n\n{}{}",
            cwd_str, allowed_dir_str, history_text, file_ops_text
        );

        ProviderResult {
            values: json!({
                "shellHistory": history_text,
                "currentWorkingDirectory": cwd_str,
                "allowedDirectory": allowed_dir_str,
            }),
            text,
            data: json!({
                "historyCount": history.len(),
                "cwd": cwd_str,
                "allowedDir": allowed_dir_str,
            }),
        }
    }
}

File: plugins/plugin-shell/python/elizaos_plugin_shell/providers/shell_history.py
from dataclasses import dataclass

from elizaos_plugin_shell.service import ShellService

MAX_OUTPUT_LENGTH = 8000
TRUNCATE_SEGMENT_LENGTH = 4000


@dataclass
class ProviderResult:
    values: dict
    text: str
    data: dict


class ShellHistoryProvider:
    @property
    def name(self) -> str:
        return "SHELL_HISTORY"

    @property
    def description(self) -> str:
        return (
            "Provides recent shell command history, current working directory, "
            "and file operations within the restricted environment"
        )

    @property
    def position(self) -> int:
        return 99

    async def get(
        self,
        message: dict,
        state: dict,
        service: ShellService | None = None,
    ) -> ProviderResult:
        if service is None:
            return ProviderResult(
                values={
                    "shellHistory": "Shell service is not available",
                    "currentWorkingDirectory": "N/A",
                    "allowedDirectory": "N/A",
                },
                text="# Shell Status\n\nShell service is not available",
                data={
                    "historyCount": 0,
                    "cwd": "N/A",
                    "allowedDir": "N/A",
                },
            )

        conversation_id = message.get("room_id") or message.get("agent_id") or "default"

        history = service.get_command_history(conversation_id, limit=10)
        cwd = service.get_current_directory(conversation_id)
        allowed_dir = service.get_allowed_directory()

        if not history:
            history_text = "No commands in history."
        else:
            history_entries = []
            for entry in history:
                entry_str = f"[{entry.timestamp}] {entry.working_directory}> {entry.command}"

                if entry.stdout:
                    stdout = entry.stdout
                    if len(stdout) > MAX_OUTPUT_LENGTH:
                        stdout = (
                            f"{stdout[:TRUNCATE_SEGMENT_LENGTH]}\n"
                            f"  ... [TRUNCATED] ...\n"
                            f"  {stdout[-TRUNCATE_SEGMENT_LENGTH:]}"
                        )
                    entry_str += f"\n  Output: {stdout}"

                if entry.stderr:
                    stderr = entry.stderr
                    if len(stderr) > MAX_OUTPUT_LENGTH:
                        stderr = (
                            f"{stderr[:TRUNCATE_SEGMENT_LENGTH]}\n"
                            f"  ... [TRUNCATED] ...\n"
                            f"  {stderr[-TRUNCATE_SEGMENT_LENGTH:]}"
                        )
                    entry_str += f"\n  Error: {stderr}"

                entry_str += f"\n  Exit Code: {entry.exit_code}"

                if entry.file_operations:
                    entry_str += "\n  File Operations:"
                    for op in entry.file_operations:
                        if op.secondary_target:
                            entry_str += f"\n    - {op.type}: {op.target} → {op.secondary_target}"
                        else:
                            entry_str += f"\n    - {op.type}: {op.target}"

                history_entries.append(entry_str)

            history_text = "\n\n".join(history_entries)

        recent_file_ops = []
        for entry in history:
            if entry.file_operations:
                recent_file_ops.extend(entry.file_operations)
        recent_file_ops = recent_file_ops[:5]

        file_ops_text = ""
        if recent_file_ops:
            ops_strs = []
            for op in recent_file_ops:
                if op.secondary_target:
                    ops_strs.append(f"- {op.type}: {op.target} → {op.secondary_target}")
                else:
                    ops_strs.append(f"- {op.type}: {op.target}")
            file_ops_text = "\n\n# Recent File Operations\n\n" + "\n".join(ops_strs)

        text = (
            f"Current Directory: {cwd}\n"
            f"Allowed Directory: {allowed_dir}\n\n"
            f"# Shell History (Last 10)\n\n{history_text}{file_ops_text}"
        )

        return ProviderResult(
            values={
                "shellHistory": history_text,
                "currentWorkingDirectory": str(cwd),
                "allowedDirectory": str(allowed_dir),
            },
            text=text,
            data={
                "historyCount": len(history),
                "cwd": str(cwd),
                "allowedDir": str(allowed_dir),
            },
        )
